store/store.go

```diff
@@ -0,0 +1,322 @@
+	r.Header.Set("Accept", "application/json")
+	client := http.Client{}
+	_, err = client.Do(r)
+	if err != nil {
+		log.Printf("Error submitting the account-key assertion: %v", err)
+		return err
+	}
+
+	// defer resp.Body.Close()
+
+	// b, err := ioutil.ReadAll(resp.Body)
+	// log.Println(string(b))
+
+	return nil
+}
+
+// LoginUser logs user in the store and returns the authentication macaroons.
+func LoginUser(username, password, otp string, permissions []string) (string, string, error) {
+	macaroon, err := requestStoreMacaroon(permissions)
+	if err != nil {
+		return "", "", err
+	}
+	deserializedMacaroon, err := auth.MacaroonDeserialize(macaroon)
+	if err != nil {
+		log.Println("Error deserializing macaroon", err)
+		return "", "", err
+	}
+
+	// get SSO 3rd party caveat, and request discharge
+	loginCaveat, err := loginCaveatID(deserializedMacaroon)
+	if err != nil {
+		log.Println("Error with login caveat", err)
+		return "", "", err
+	}
+
+	discharge, err := dischargeAuthCaveat(loginCaveat, username, password, otp)
+	if err != nil {
+		log.Println("Error with discharge", err)
+		return "", "", err
+	}
+
+	return macaroon, discharge, nil
+}
+
+func generateAccountKeyRequest(keyAuth KeyRegister, keypair datastore.Keypair) (string, error) {
+	// Create a self-signed account-key assertion
+	since := time.Now().AddDate(-1, 0, 0)
+	headers := map[string]interface{}{
+		"account-id":          keyAuth.AuthorityID,
+		"name":                keyAuth.KeyName,
+		"public-key-sha3-384": keypair.KeyID,
+		"since":               since.Format(time.RFC3339),
+	}
+
+	// Loads the keypair into the memory keystore
+	err := datastore.Environ.KeypairDB.LoadKeypair(keypair.AuthorityID, keypair.KeyID, keypair.SealedKey)
+	if err != nil {
+		log.Println("Error loading the keypair", err)
+		return "", err
+	}
+
+	// Get the public key as it is the body of the assertion
+	publicKey, err := datastore.Environ.KeypairDB.PublicKey(keypair.KeyID)
+	if err != nil {
+		log.Println("Error fetching the public key", err)
+		return "", err
+	}
+	pubKeyEncoded, err := asserts.EncodePublicKey(publicKey)
+	if err != nil {
+		log.Println("Error encoding the public key", err)
+		return "", err
+	}
+
+	accountKey, err := datastore.Environ.KeypairDB.SignAssertion(asserts.AccountKeyRequestType, headers, pubKeyEncoded, keypair.AuthorityID, keypair.KeyID, keypair.SealedKey)
+	if err != nil {
+		log.Printf("Error creating account-key assertion: %v", err)
+		return "", err
+	}
+
+	assert := asserts.Encode(accountKey)
+	return string(assert), nil
+}
+
+func requestStoreMacaroon(permissions []string) (string, error) {
+
+	perm := Permissions{Permissions: permissions}
+	macaroonJSONData, err := json.Marshal(perm)
+	if err != nil {
+		log.Println("Error marshalling the macaroon", err)
+		return "", err
+	}
+
+	// Submit the account-key assertion
+	r, err := http.Post(storeBaseURL+"acl/", "application/json", bytes.NewReader(macaroonJSONData))
```

Comment:
same as before, maybe it would be nice having all http client operations in same file/place

Reply:
I'll look at it, though the style of the two calls is different.

Reply:
I've done it, but it doesn't look great to me. They are different calls: one with custom headers and one without, one with a straight POST and the other with the response being parsed.